File: frontend/src/engine/capabilities/transpose.ts
/**
 * transpose – Flip rows and columns of a range.
 *
 * Reads the source range, transposes the 2D array, writes to outputRange.
 * Values-only by default (formatting not copied).
 */

import { CapabilityMeta, TransposeParams, StepResult, ExecutionOptions } from "../types";
import { registry } from "../capabilityRegistry";
import { resolveRange } from "./rangeUtils";

const meta: CapabilityMeta = {
  action: "transpose",
  description: "Transpose a range — flip rows and columns",
  mutates: true,
  affectsFormatting: false,
};

async function handler(
  context: Excel.RequestContext,
  params: TransposeParams,
  options: ExecutionOptions,
): Promise<StepResult> {
  const { sourceRange, outputRange } = params;

  if (options.dryRun) {
    return { stepId: "", status: "success", message: `Would transpose ${sourceRange} → ${outputRange}` };
  }

  options.onProgress?.("Reading source range...");
  const srcRng = resolveRange(context, sourceRange);
  srcRng.load("values, rowCount, columnCount");
  await context.sync();

  const vals = (srcRng.values ?? []) as (string | number | boolean | null)[][];
  if (!vals.length) return { stepId: "", status: "success", message: "No data to transpose." };

  const rows = vals.length;
  const cols = vals[0]?.length ?? 0;
  options.onProgress?.(`Transposing ${rows} × ${cols}...`);

  // Transpose: out[c][r] = vals[r][c]
  const transposed: (string | number | boolean | null)[][] = Array.from({ length: cols }, (_, c) =>
    Array.from({ length: rows }, (_, r) => vals[r]?.[c] ?? null),
  );

  const outRng = resolveRange(context, outputRange);
  outRng.getResizedRange(transposed.length - 1, (transposed[0]?.length ?? 1) - 1).values = transposed as any;
  await context.sync();

  return {
    stepId: "",
    status: "success",
    message: `Transposed ${rows}×${cols} → ${cols}×${rows} written to ${outputRange}`,
    outputs: { outputRange },
  };
}

registry.register(meta, handler as any);
export { meta };

File: backend/app/execution/capabilities/transpose.py
"""transpose — flip rows and columns of a range.

Reads the source range, transposes the 2D array, and writes the result to the
output range. Values-only by default (formatting is not copied).
"""

from __future__ import annotations

from typing import Any

from app.execution.base import ExecutorContext
from app.execution.capability_registry import registry
from app.execution.range_utils import resolve_range


def handler(ctx: ExecutorContext, params: dict[str, Any]) -> dict[str, Any]:
    source = params.get("sourceRange")
    output = params.get("outputRange")
    if not source or not output:
        return {
            "status": "error",
            "message": "transpose requires 'sourceRange' and 'outputRange'.",
        }

    if ctx.dry_run:
        return {
            "status": "preview",
            "message": f"Would transpose {source} → {output}.",
        }

    try:
        src = resolve_range(ctx.workbook_handle, source)
        raw = src.value

        # Normalize to 2D matrix — xlwings returns scalar for 1 cell, list for
        # a single row/column, 2D list otherwise. Match the TS behavior.
        if raw is None:
            vals: list[list[Any]] = []
        elif not isinstance(raw, list):
            vals = [[raw]]
        elif raw and not isinstance(raw[0], list):
            # Single row or single column — disambiguate by rng shape.
            rows, cols = src.shape
            if rows == 1:
                vals = [list(raw)]
            else:
                vals = [[v] for v in raw]
        else:
            vals = [list(r) for r in raw]

        if not vals:
            return {"status": "success", "message": "No data to transpose.", "outputs": {}}

        rows = len(vals)
        cols = max((len(r) for r in vals), default=0)

        # Transpose: out[c][r] = vals[r][c]
        transposed: list[list[Any]] = [
            [vals[r][c] if c < len(vals[r]) else None for r in range(rows)]
            for c in range(cols)
        ]

        out_rows = len(transposed)
        out_cols = len(transposed[0]) if transposed else 0
        out_rng = resolve_range(ctx.workbook_handle, output)
        out_rng.resize(out_rows, out_cols).value = transposed
    except Exception as exc:  # noqa: BLE001
        return {"status": "error", "message": f"Transpose failed: {exc}", "error": str(exc)}

    return {
        "status": "success",
        "message": f"Transposed {rows}×{cols} → {out_rows}×{out_cols} written to {output}.",
        "outputs": {"outputRange": output},
    }


registry.register("transpose", handler, mutates=True)
